src/gluonnlp/model/bert.py

```diff
@@ -30,13 +30,225 @@
+        with self.name_scope():
+            if self._use_bias:
+                self.query_bias = self.params.get('query_bias', shape=(self.units,),
+                                                  init=bias_initializer)
+                self.key_bias = self.params.get('key_bias', shape=(self.units,),
+                                                init=bias_initializer)
+                self.value_bias = self.params.get('value_bias', shape=(self.units,),
+                                                  init=bias_initializer)
+            weight_shape = (self.units, self.units)
+            self.query_weight = self.params.get('query_weight', shape=weight_shape,
+                                                init=weight_initializer,
+                                                allow_deferred_init=True)
+            self.key_weight = self.params.get('key_weight', shape=weight_shape,
+                                              init=weight_initializer,
+                                              allow_deferred_init=True)
+            self.value_weight = self.params.get('value_weight', shape=weight_shape,
+                                                init=weight_initializer,
+                                                allow_deferred_init=True)
+            self.dropout_layer = nn.Dropout(self._dropout)
+
+    def _collect_params_with_prefix(self, prefix=''):
+        # the registered parameter names in v0.8 are the following:
+        # prefix_proj_query.weight, prefix_proj_query.bias
+        # prefix_proj_value.weight, prefix_proj_value.bias
+        # prefix_proj_key.weight, prefix_proj_key.bias
+        # this is a temporary fix to keep backward compatibility, due to an issue in MXNet:
+        # https://github.com/apache/incubator-mxnet/issues/17220
+        if prefix:
+            prefix += '.'
+        ret = {prefix + 'proj_' + k.replace('_', '.') : v for k, v in self._reg_params.items()}
+        for name, child in self._children.items():
+            ret.update(child._collect_params_with_prefix(prefix + name))
+        return ret
+
+    # pylint: disable=arguments-differ
+    def hybrid_forward(self, F, qkv, valid_len, query_bias, key_bias, value_bias,
+                       query_weight, key_weight, value_weight):
+        # interleaved_matmul_selfatt ops assume the projection is done with interleaving
+        # weights for query/key/value. The concatenated weight should have shape
+        # (num_heads, C_out/num_heads * 3, C_in).
+        query_weight = query_weight.reshape(shape=(self._num_heads, -1, 0), reverse=True)
+        key_weight = key_weight.reshape(shape=(self._num_heads, -1, 0), reverse=True)
+        value_weight = value_weight.reshape(shape=(self._num_heads, -1, 0), reverse=True)
+        in_weight = F.concat(query_weight, key_weight, value_weight, dim=-2)
+        in_weight = in_weight.reshape(shape=(-1, 0), reverse=True)
+        in_bias = F.concat(query_bias, key_bias, value_bias, dim=0)
```

Comment:
Is it possible to avoid concat for every iteration? Or at least for inference, we only need concat once, right?

Reply:
For inference, yes that's true. It's similar to RNN. If we figure out a way to avoid the weight concat in RNN, we can apply that here, too. @TaoLv do you have any idea/suggestion?